tests/transforms/test_compile.py

```diff
@@ -0,0 +1,513 @@
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import pytest
+
+from pennylane import numpy as np
+
+import pennylane as qml
+from pennylane.wires import Wires
+
+from pennylane.transforms.optimization import cancel_inverses, commute_controlled, merge_rotations
+
+from test_optimization.utils import compare_operation_lists
+
+
+def build_qfunc(wires):
+    def qfunc(x, y, z):
+        qml.Hadamard(wires=wires[0])
+        qml.RZ(z, wires=wires[2])
+        qml.CNOT(wires=[wires[2], wires[1]])
+        qml.CNOT(wires=[wires[1], wires[0]])
+        qml.RX(x, wires=wires[0])
+        qml.CNOT(wires=[wires[1], wires[0]])
+        qml.RZ(-z, wires=wires[2])
+        qml.RX(y, wires=wires[0])
+        qml.PauliY(wires=wires[2])
+        qml.CY(wires=[wires[1], wires[2]])
+        return qml.expval(qml.PauliZ(wires=wires[0]))
+
+    return qfunc
+
+
+class TestCompile:
+    """Unit tests for compilation."""
+
+    def test_invalid_pipeline(self):
+        """Test that error is raised for an invalid function in the pipeline"""
+        qfunc = build_qfunc([0, 1, 2])
+        dev = qml.device("default.qubit", wires=[0, 1, 2])
+
+        transformed_qfunc = qml.compile(pipeline=[cancel_inverses, isinstance])(qfunc)
+        transformed_qnode = qml.QNode(transformed_qfunc, dev)
+
+        with pytest.raises(ValueError, match="Invalid transform function"):
+            transformed_qnode(0.1, 0.2, 0.3)
+
+    def test_invalid_basis_gates(self):
+        """Test that error is raised for an invalid basis gate."""
+        qfunc = build_qfunc([0, 1, 2])
+        dev = qml.device("default.qubit", wires=[0, 1, 2])
+
+        transformed_qfunc = qml.compile(basis_set=["CNOT", "RX", "RY", "RZ", "Q2"])(qfunc)
+        transformed_qnode = qml.QNode(transformed_qfunc, dev)
+
+        with pytest.raises(ValueError, match="Invalid basis gate"):
+            transformed_qnode(0.1, 0.2, 0.3)
```

Comment:
Seeing this as a unit test reinforces to me that we probably shouldn't be validating the users basis gates?